ext/standard/streamsfuncs.c

```diff
@@ -756,115 +810,150 @@ static int stream_array_emulate_read_fd_set(zval *stream_array)
 /* }}} */
 
 /* {{{ Runs the select() system call on the sets of streams with a timeout specified by tv_sec and tv_usec */
-PHP_FUNCTION(stream_select)
-{
-	zval *r_array, *w_array, *e_array;
-	struct timeval tv, *tv_p = NULL;
-	fd_set rfds, wfds, efds;
-	php_socket_t max_fd = 0;
-	int retval, sets = 0;
-	zend_long sec, usec = 0;
-	bool secnull;
-	bool usecnull = 1;
-	int set_count, max_set_count = 0;
-
-	ZEND_PARSE_PARAMETERS_START(4, 5)
-		Z_PARAM_ARRAY_EX2(r_array, 1, 1, 0)
-		Z_PARAM_ARRAY_EX2(w_array, 1, 1, 0)
-		Z_PARAM_ARRAY_EX2(e_array, 1, 1, 0)
-		Z_PARAM_LONG_OR_NULL(sec, secnull)
-		Z_PARAM_OPTIONAL
-		Z_PARAM_LONG_OR_NULL(usec, usecnull)
-	ZEND_PARSE_PARAMETERS_END();
+PHP_FUNCTION(stream_select) {
+    zval *r_array, *w_array, *e_array;
+    struct timeval tv, *tv_p = NULL;
+    fd_bigset rfds, wfds, efds;
+    php_socket_t max_fd = 0;
+    int retval, sets = 0;
+    zend_long sec, usec = 0;
+    bool secnull;
+    bool usecnull = 1;
+    int set_count = 0, max_set_count = 0;
+
+    ZEND_PARSE_PARAMETERS_START(4, 5)
+        Z_PARAM_ARRAY_EX2(r_array, 1, 1, 0)
+        Z_PARAM_ARRAY_EX2(w_array, 1, 1, 0)
+        Z_PARAM_ARRAY_EX2(e_array, 1, 1, 0)
+        Z_PARAM_LONG_OR_NULL(sec, secnull)
+        Z_PARAM_OPTIONAL
+        Z_PARAM_LONG_OR_NULL(usec, usecnull)
+    ZEND_PARSE_PARAMETERS_END();
+
+#ifdef PHP_WIN32
+	/* Check if the Windows standard handle is redirected to file */
+	long max_fds = FD_SETSIZE;
+#else
+	/* Check if the file descriptor identifier is a terminal */
+	long max_fds = sysconf(_SC_OPEN_MAX);
+	if (max_fds == -1) {
+		max_fds = FD_SETSIZE;
+	}
+#endif
```

Comment:
`_SC_OPEN_MAX` can theoretically be as high as INT_MAX. On my system, Docker starts any container with `RLIMIT_NOFILE` set to `1073741816` by default, so we would allocate 127MiB per fd set here, 381MiB in total.

An alternative to this would be to grow the sets on the fly in `FD_BIGSET_SET`:

```
void fd_bitset_grow(fd_bigset *set, int fd) {
    ...
}

#define FD_BIGSET_SET(fd, set) do { \
    if (UNEXPECTED((fd) * 8 > (set)->size) { \
        fd_bitset_grow(set, fd);
    }
    (set)->fds_bits[(fd) / 8] |= (1 << ((fd) % 8)); \
} while (0)
```

In fd_bitset_grow(), take the max of `fd/8` and `set->size*2` (to limit the number of reallocs) and round up to the next power of two.

Then before calling select() we can adjust the size of the 3 sets so they have the same size.

Reply:
How about just setting an upper bound at for example 128k descriptors unless fd_setsize is larger? Possibly in combination with your doubling approach?

Reply:
At the minimum we need a reallocation scheme so we can support all cases. If we start with `FD_SETSIZE`, we get no performance regression for the cases that already work today.

We can use a bigger start size as you suggest, but I'm not sure it will be a win. It may incur a regression for the common case of a few small fds. Ideally we should measure it with different variations of number of file descriptors and highest descriptor numbers.

Reply:
@arnaud-lb I don't think the difference between allocating 16 kB instead of 1 kB is at all significant in the scope of the entire function invocation. Generally, allocating 1 kB means allocating 4 kB (page size). This function is not part of a very busy loop either; it is designed to sleep until there is activity on one of the file descriptors. I would not call this a regression, since it for a tiny cost in temporary memory allocation means that the FD_SETSIZE but fixes an fairly undocumented problem that has been in PHP for a decade.

Reply:
> This function is not part of a very busy loop either; it is designed to sleep until there is activity on one of the file descriptors.

I have an async job system where this function is very much on a hot path, fwiw.

Reply:
@withinboredom Generally, the job system should perform all non-blocking tasks and then invoke this function once, before resuming to perform all those non-blocking tasks. So what I mean is that this function is a tiny fraction of all the work involved in one iteration of an async event loop. If the loop is empty (no CPU bound work to do), the stream_select function should be invoked with a small timeout to avoid busy waiting and allowing other server processes to work.

Reply:
I just pushed a fixed file where editorconfig is added

Reply:
I fully support this change, but I think we can make it better without too much effort. I can take care of that if you don't feel like it.

If we are not sure about the performance impact we should measure it, but here is my intuition: The allocator uses different strategies depending on block size, and is considerably faster for smaller blocks (below 4kB) as it's usually a [fast O(1) operation](https://github.com/php/php-src/blob/df6d85acf852e085318293b1bcbf9f9d384e5bea/Zend/zend_alloc.c#L1447-L1450). For larger blocks it's a non-linear algorithm and it's subject to fragmentation. So allocating 1kB (or 128 bytes for FD_SETSIZE=1024) will be considerably faster than allocating 16kB. Moreover, this memory is zeroed after allocation, which also has a cost.

Reply:
@arnaud-lb  It is much more complex than you realize to scale the size of the bitmap. My first implementation did exactly this. The tests revealed that user space stream wrappers emit notifications that can't be prevented. So you can only scan the array and cast it to file descriptors once, or else two warnings will be emitted and tests will fail. So you need to allocate a temporary buffer to cache the file descriptor ids first, so that you can iterate over the list twice. First to determine what is the largest file descriptor id, and then after allocating a suitable sized bitmap, then you need to iterate it again to set the appropriate bits. For 10000 or more file descriptors, the cost of this will be quite large, very likely costing a lot more than the microsecond or so it costs to allocate 16 kb instead of 1 kb.

Reply:
Why can't the engine just keep track of the largest file descriptor it has opened so far?

Reply:
@withinboredom I like the suggestion. I am not sure if we can be sure we cover all ways a file descriptor is created. File descriptors can also be received, and we might get som thread safety issues for ZTS?

Reply:
I ran a quick benchmark, and as you predicted @frodeborli there is no performance degradation associated with always allocating a set for 131072 fds, at least in this benchmark.

However, I think that having a hard limit is unfortunate, as this only half-fixes the issue. Of course it's possible to re-compile with a higher limit, but that's not great.

Now I like the suggestion of @withinboredom, but this may be an intrusive change. As an alternative, we could remember the max fd that stream_select() has seen, and use that for the initial allocation size. We could even average this max fd over time (with just two points of data), so that a temporary increase of fd size only has a temporary impact. Then, if necessary, re-alloc on the fly when iterating over the fds, as I suggested in https://github.com/php/php-src/pull/14452#discussion_r1633168339 (only a single pass on the fds is necessary).

Reply:
With my latest commit, I'm dynamically increasing the bitmap size via the FD_BIGSET_SET macro if a file descriptor >=131072 is set. The FD_SETSIZE configure option may still be relevant since PHP uses FD_SET internally other places in the code. But for userspace code, it's no longer relevant with this commit AFAIK.

Reply:
Thank you! I did a new pass on the code and I think we are almost good (apart from a few comments).

Could you add a few tests? [This](https://github.com/php/php-src/blob/b229f189ee22906da59c5b2502a1aa38bc66ff3e/ext/standard/tests/streams/stream_select_preserve_keys.phpt) is an example of a test using stream_select(). You may need to introduce a zend_test_dup2() function as suggested in https://github.com/php/php-src/pull/14452#discussion_r1624626025. You can then convert a fd to a stream with the php://fd/ wrapper: https://www.php.net/manual/en/wrappers.php.php.